### PR TITLE
Add support for file system globbing of project paths.

### PR DESCRIPTION
**Pull request by geeknoid**
With this change, you can now do:

slngen **\*.csproj

And get a solution for all the projects in a directory hierarchy. If
you expand the command-line to:

slngen --folders true --collapsefolders true --ignoreMainProject **\*.csproj

You get a pretty nice-looking solution file that is close to a hand-crafted
one.

### CODE REVIEW

File: src/Microsoft.VisualStudio.SlnGen.Tool/Microsoft.VisualStudio.SlnGen.Tool.csproj [hunk 1]
```diff
@@ -13,4 +13,7 @@
   <ItemGroup>
     <FilesToSign Include="$(IntermediateOutputPath)$(TargetName)$(TargetExt)" Authenticode="Microsoft400" StrongName="StrongName" />
   </ItemGroup>
+  <ItemGroup>
```
Comment (psfinaki) [inline src/Microsoft.VisualStudio.SlnGen.Tool/Microsoft.VisualStudio.SlnGen.Tool.csproj:16]: Looks like this way it cannot find the package in `ProgramArguments.Shared.cs` and fails the build :(

Comment (jeffkl) [reply src/Microsoft.VisualStudio.SlnGen.Tool/Microsoft.VisualStudio.SlnGen.Tool.csproj:16]: Add the PackageReference to [Shared.props](https://github.com/microsoft/slngen/blob/master/src/Shared/Shared.props#L16) which is imported by both projects.  We ship a .NET Core global tool and a .NET Framework console app.

Comment (geeknoid) [reply src/Microsoft.VisualStudio.SlnGen.Tool/Microsoft.VisualStudio.SlnGen.Tool.csproj:16]: OK, thanks for the pointer. I've updated accordingly.

File: Packages.props [hunk 2]
```diff
@@ -18,6 +18,7 @@
     <PackageReference Update="Shouldly" Version="4.0.3" />
     <PackageReference Update="xunit" Version="2.4.1" />
     <PackageReference Update="xunit.runner.visualstudio" Version="2.4.3" />
+    <PackageReference Update="Microsoft.Extensions.FileSystemGlobbing" Version="5.0.0" />
```
Comment (jeffkl) [inline Packages.props:21]: Please add this in sorted order

File: src/Shared/ProgramArguments.Shared.cs [hunk 3]
```diff
@@ -387,6 +392,37 @@ public bool TryGetEntryProjectPaths(ISlnGenLogger logger, out IReadOnlyList<stri
             return result.Count > 0;
         }
 
+        private static IEnumerable<string> ExpandWildcards(IEnumerable<string> paths)
+        {
+#if NETFRAMEWORK
+            return paths;
```
Comment (jeffkl) [inline src/Shared/ProgramArguments.Shared.cs:398]: Why can't this work for .NET Framework?

File: src/Shared/ProgramArguments.Shared.cs [hunk 4]
```diff
@@ -387,6 +392,37 @@ public bool TryGetEntryProjectPaths(ISlnGenLogger logger, out IReadOnlyList<stri
             return result.Count > 0;
         }
 
+        private static IEnumerable<string> ExpandWildcards(IEnumerable<string> paths)
```
Comment (jeffkl) [inline src/Shared/ProgramArguments.Shared.cs:395]: Please add a unit test for this method to ensure it doesn't get regressed

File: src/Shared/ProgramArguments.Shared.cs (hunk 3 above)
Comment (geeknoid) [reply src/Shared/ProgramArguments.Shared.cs:398]: The globbing library depends on newer underlying runtime.

Comment (geeknoid) [reply src/Shared/ProgramArguments.Shared.cs:398]: Actually, the issue is that the globbing library needs either .NET 4.61, .NET 4.72, or .NET Standard 2.0. Right now, SlnGen is being built against .NET 4.6 and 4.71 and the globbing library won't tolerate.

So I can either leave the new feature so it only works for .NET Core versions, or I can increase the framework versions that the rest of SlnGen is built with.

File: src/Shared/ProgramArguments.Shared.cs (hunk 4 above)
Comment (geeknoid) [reply src/Shared/ProgramArguments.Shared.cs]: Done

File: Packages.props (hunk 2 above)
Comment (geeknoid) [reply Packages.props:21]: Done.

**Comment by geeknoid** [comment]
So here's the current state:

- The new globbing functionality only works in the .NET Core client, not in the .NET Framework client. This is because the globbing library doesn't want to bind to 4.6, only to 4.61. When I change the few SlnGen projects to build with this framework, then other dependencies suddenly break and demand higher versions. So I left things alone and just made the functionality run on the newer clients.

- The unit tests were previously only build for net472. In order to be able to test the new functionality, which is only in .NET Core versions, I made the unit tests so they are also built and run against .NET 5.0. This mostly just worked, except for a few specific unit tests that broke in that mode. I conditionalize those tests so they only run in the 4.72 version. You might want to dig deeper and figure out why those tests are failing on .NET 5.0.

Thanks.

File: src/Microsoft.VisualStudio.SlnGen.UnitTests/ArgumentsTests.cs [hunk 5]
```diff
@@ -57,5 +59,51 @@ public void NoLogoHidesLogo(string argument)
...
+                File.WriteAllText(Path.Combine(work.FullName, "t1", "t2", "4.csproj"), string.Empty);
+                File.WriteAllText(Path.Combine(work.FullName, "t1", "t2", "5.csproj"), string.Empty);
+                File.WriteAllText(Path.Combine(work.FullName, "t1", "t2", "t3", "6.csproj"), string.Empty);
+
+                var old = Environment.CurrentDirectory;
+                Environment.CurrentDirectory = work.FullName;
+                var result = ProgramArguments.ExpandWildcards(new[] { "**\\*.csproj" });
+                Environment.CurrentDirectory = old;
+
+                var files = result.Select(x => Path.GetFileName(x)).ToArray();
```
Comment (psfinaki) [inline src/Microsoft.VisualStudio.SlnGen.UnitTests/ArgumentsTests.cs:93]: `OrderBy()` instead?

Comment (geeknoid) [reply src/Microsoft.VisualStudio.SlnGen.UnitTests/ArgumentsTests.cs]: Done.

**Review by psfinaki — APPROVED** [review]
Approve from me - but wait for maintainers :D (I just like this project and follow it on GitHub)

**Review by jeffkl — APPROVED** [review]
Thanks for the contribution.  I'm going to merge this and then look into a way to have it work for .NET Framework.

**Pull request merged**